middleware/rewrite/edns0.go

```diff
@@ -0,0 +1,88 @@
+// Package rewrite is middleware for rewriting requests internally to something different.
+package rewrite
+
+import (
+	"encoding/hex"
+	"log"
+	"strconv"
+
+	"github.com/miekg/dns"
+)
+
+// Edns0Rule is a class rewrite rule.
+type Edns0Rule struct {
+	action string
+	code   uint16
+	data   []byte
+	match  []byte // TODO: add match input for replace rule
+}
+
+// SetEDNS0Attrs will alter the request
+func (rule *Edns0Rule) SetEDNS0Attrs(r *dns.Msg) {
+	o := r.IsEdns0()
+	if o == nil {
+		return
```

Comment:
If we are doing append or set, then I would think we could be adding an EDNS0 opt record to a Msg that doesn't have one. So, I don't think we want to make this test here. Instead, we just want to skip the for loop below if o is nil.